tests/auth_tests/test_models.py

```diff
@@ -261,6 +261,172 @@ def test_check_password_upgrade(self):
+        )
+
+    @override_settings(AUTHENTICATION_BACKENDS=['auth_tests.test_auth_backends.CustomModelBackend'])
+    def test_custom_backend_with_permission_instance_and_obj(self):
+        obj = CustomModel.objects.create(user=self.charlie_brown)
+        self.assertCountEqual(
+            User.objects.with_perm(self.permission, obj=obj),
+            [self.charlie_brown],
+        )
+
+    def test_invalid_perm_type(self):
+        msg = 'The `perm` argument must be a string or a permission instance.'
+
+        class FakePermission:
+            pass
+
+        for perm in (b'auth.test', FakePermission()):
+            with self.subTest(perm=perm):
+                with self.assertRaisesMessage(TypeError, msg):
+                    User.objects.with_perm(perm)
+
+    @override_settings(AUTHENTICATION_BACKENDS=[
+        'auth_tests.test_auth_backends.CustomModelBackend',
+        'django.contrib.auth.backends.ModelBackend',
+    ])
+    def test_custom_backend(self):
+        msg = (
+            'You have multiple authentication backends configured and '
+            'therefore must provide the `backend` argument.'
+        )
+        backend = 'auth_tests.test_auth_backends.CustomModelBackend'
+        with self.assertRaisesMessage(ValueError, msg):
+            User.objects.with_perm('auth.test')
+        self.assertCountEqual(
+            User.objects.with_perm('auth.test', backend=backend),
+            [self.charlie, self.charlie_brown],
+        )
+        self.assertNotIn(self.user1, User.objects.with_perm('auth.test', backend=backend))
+
+    def test_default_backend_with_obj(self):
+        obj = CustomModel.objects.create(user=self.charlie_brown)
+        self.assertCountEqual(
+            User.objects.with_perm('auth.test', obj=obj),
+            [],
+        )
+
+    @override_settings(AUTHENTICATION_BACKENDS=['auth_tests.test_auth_backends.CustomModelBackend'])
+    def test_custom_backend_with_obj(self):
+        obj = CustomModel.objects.create(user=self.charlie_brown)
+        self.assertCountEqual(
+            User.objects.with_perm('auth.test', obj=obj),
+            [self.charlie_brown],
+        )
+
+    def test_invalid_backend_type(self):
+        msg = 'The `backend` argument must be a string.'
+        with self.assertRaisesMessage(TypeError, msg):
+            User.objects.with_perm(
+                'auth.test',
+                backend=b'auth_tests.test_auth_backends.CustomModelBackend',
+            )
+
+    def test_nonexistent_backend(self):
+        with self.assertRaises(ImportError):
+            User.objects.with_perm(
+                'auth.test',
+                backend='invalid.backend.CustomModelBackend',
+            )
+
+    def test_invalid_permission(self):
+        self.assertCountEqual(
+            User.objects.with_perm('invalid.perm'),
+            [self.superuser],
+        )
+
+    @override_settings(AUTHENTICATION_BACKENDS=['auth_tests.test_auth_backends.BareModelBackend'])
```

Comment:
You can use built-in `django.contrib.auth.backends.BaseBackend` instead of custom `BareModelBackend`.